### ISSUE TITLE
TSM engine breaks windows compatibility

### ISSUE THREAD

**Issue by corylanou** [influxdb/influxdb#4358]
```
$ GOOS=windows GOARCH=amd64 go build ./cmd/influxd
# github.com/influxdb/influxdb/tsdb/engine/tsm1
tsdb/engine/tsm1/tsm1.go:1799: undefined: syscall.Mmap
tsdb/engine/tsm1/tsm1.go:1799: undefined: syscall.PROT_READ
tsdb/engine/tsm1/tsm1.go:1799: undefined: syscall.MAP_SHARED
tsdb/engine/tsm1/tsm1.go:1836: undefined: syscall.Munmap
```

@pauldix @jwilder

**Comment by corylanou** [comment]
@benbjohnson already handles this with boltdb and we should be able to do the same thing:

https://github.com/boltdb/bolt/blob/master/bolt_windows.go#L30

**Comment by otoolep** [comment]
+1, sounds like a good idea.

**Comment by edokan** [comment]
I've tried mmap-go package, and it seems to be working.

ref: https://github.com/HouzuoGuo/tiedot/issues/7

**Comment by mvadu** [comment]
Any ETA on the fix for this issue? Until its fixed, I have to stay in old builds!

**Comment by imrichardcole** [comment]
What is the latest version of influxdb that will compile on Windows without hitting this issue please?

**Comment by otoolep** [comment]
@imrichardcole -- check out the 0.9.4 branch and compile that. It should work on Windows.

**Comment by imrichardcole** [comment]
@otoolep Ah many thanks, you'd actually responded before I had to chance to post back.  I managed to get 0.9.4.2 to compile on Windows.

Can we expect ongoing Windows support or is it limited to 0.9.4.2?

**Comment by otoolep** [comment]
@imrichardcole -- our Windows support is currently experimental, and the only guarantee we can provide is that when 1.0 comes out it will run on Windows.

**Comment by aadrian** [comment]
@otoolep the windows version is used mostly for development, not for production, so please release it even if it's experimental.

Thank you.

**Comment by corylanou** [comment]
We are hoping to add experimental support back soon.  This PR will hopefully allow for it if it gets merged for `0.9.6`: https://github.com/influxdb/influxdb/pull/4788

Any help from the community will get this done sooner.

**Comment by mvadu** [comment]
Until that time, I just built and published a release from 0.9.4 branch, https://github.com/mvadu/influxdb/releases/tag/v0.9.4.2

**Comment by mvadu** [comment]
@corylanou @otoolep p  I tried adding mmap implementation for Windows based on MapViewOfFile. Used SliceHeader trick to change the pointer returned by MapViewOfFile to a byte slice. This will not call for any change in rest of tsm.

Please see the changes in https://github.com/mvadu/influxdb/pull/1 

However I am not sure where this mmap function is called, as go build is still complains about 
tsdb\engine\tsm1\tsm1.go:1974: undefined: syscall.Mmap
tsdb\engine\tsm1\tsm1.go:1974: undefined: syscall.PROT_READ
tsdb\engine\tsm1\tsm1.go:1974: undefined: syscall.MAP_SHARED
tsdb\engine\tsm1\tsm1.go:2033: undefined: syscall.Munmap

Note: This is the first time I am working on a fork repo. Please excuse me if I did some silly mistake. I am not sure if I should have committed on my fork's master and raise a PR to you guys.

**Comment by rapport** [comment]
Is this also related https://github.com/influxdb/influxdb/issues/4787 ?

**Comment by pauldix** [comment]
I think this is now fixed in master. Can someone confirm? Bill Gates banned me from running Windows about a decade ago.

**Comment by rapport** [comment]
It is not working for me but does appear to have less errors

ubuntu@NUC001:~/.gvm/pkgsets/go1.5.2/global/src/github.com/influxdb/influxdb$ git pull
Already up-to-date.
ubuntu@NUC001:~/.gvm/pkgsets/go1.5.2/global/src/github.com/influxdb/influxdb$ go get -u -f ./...
ubuntu@NUC001:~/.gvm/pkgsets/go1.5.2/global/src/github.com/influxdb/influxdb$ GOOS=windows GOARCH=386 go build ./...
# github.com/influxdb/influxdb/tsdb/engine/tsm1
tsdb/engine/tsm1/mmap_windows.go:13: too many arguments to return

**Comment by fazalmajid** [comment]
@rapport no, #4787 is not related, tsm1 already broke Windows, but my first pull request #4788 made things worse. The second pull request #5027 fixes #4787 and doesn't introduce any additional regressions on Windows, but sadly it does nothing to fix #4358 either.

Apparently CircleCI (used by InfluxDB) has beta support for Windows, but it looks like an Enterprise (i.e. paid) feature: https://circleci.com/enterprise/azure 

I wonder if the Microsoft Azure Open Source outreach team could be convinced to pick up the tab, but I have no idea how to get this done - my few contacts at Microsoft are at their Mobile division, not Azure.

**Comment by mvadu** [comment]
@pauldix @fazalmajid master just has `func mmap(f *os.File, offset int64, length int) ([]byte, error) { 
return nil, fmt.Errorf("mmap file not supported windows")` so not really a working version.

I tried with #4981 and its work in progress now. I think I know the reason for the pending test failures, and should be able to fix in a day.

**Comment by mvadu** [comment]
@pauldix I have fixed this issue temporarily with #5070, but it will not allow tsm1 to run in Windows yet.

**Comment by fazalmajid** [comment]
As an alternative to CircleCI, Appveyor offers free Windows CI for open-source projects: http://www.appveyor.com/
